clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -1061,6 +1127,37 @@ public void close() {
         }
     }
 
+    private static class DelayedChannelClose {
+        private final Time time;
+        private final KafkaChannel channel;
+        private final CloseMode closeMode;
+        private final long endTime;
+
+        public DelayedChannelClose(Time time, long startMs, int minimumDelayMs, KafkaChannel channel, CloseMode closeMode) {
+            this.time = time;
+            this.channel = channel;
+            this.closeMode = closeMode;
+            this.endTime = computeDelay(time, startMs, minimumDelayMs);
+        }
+
+        private long computeDelay(Time time, long startMs, int minimumDelayMs) {
+            if (startMs < 0)
+                throw new IllegalArgumentException();
+
+            long now = time.milliseconds();
+            long elapsed = now - startMs;
+            long delayMs = minimumDelayMs;
+            long remainingMs;
+            while ((remainingMs = delayMs - elapsed) < 0)
+                delayMs *= 2;
```

Comment:
This doesn't seem to match the doc for the new config? Couldn't we just delay by `connection.failed.authentication.delay.ms`?

Reply:
I tried explaining the approach in response to another comment. If you think this is reasonable, I can add code comments to explain why we do this and also update the KIP if required.